pdoc/web.py

```diff
@@ -97,24 +97,54 @@ def log_request(
 class DocServer(http.server.HTTPServer):
     """pdoc's live-reloading web server"""
 
-    all_modules: Collection[str]
+    all_modules: AllModules
 
-    def __init__(self, addr: tuple[str, int], all_modules: Collection[str], **kwargs):
+    def __init__(self, addr: tuple[str, int], specs: list[str], **kwargs):
         super().__init__(addr, DocHandler, **kwargs)  # type: ignore
-        self.all_modules = all_modules
+        module_names = extract.walk_specs(specs)
+        self.all_modules = AllModules(module_names)
 
     @cache
     def render_search_index(self) -> str:
         """Render the search index. For performance reasons this is always cached."""
-        all_mods = {}
+        # Some modules may not be importable, which means that they would raise an RuntimeError
```

Comment:
if I understand correctly this is exactly what https://docs.python.org/3/library/importlib.html#importlib.abc.MetaPathFinder.find_spec was designed for

Reply:
Not really, things are a bit more complicated here. We use `find_spec` and friends in `pdoc.extract` to get a list of all modules and submodules, but that does not guarantee yet that all those submodules are actually importable. If you module looks like [this](https://github.com/mitmproxy/pdoc/blob/ad32abc7a5e337e8ab424d173638ee516713151b/test/testdata/import_err_simple.py) `find_spec` will happily take it, but you cannot import it later on.

When pdoc is running as web server, we can do two things:

1. Import all modules on startup, as we would when running with `-o`. The first downside is that this can be a bit of a performance hog and pdoc may need multiple seconds to start up. The second downside is that pdoc picks up _all_ submodules, including platform specific ones (think _windows, _linux, ...). So for some projects you will be greeted with quite a few funny error messages, which isn't ideal either.
2. Only import modules on-demand. The upside is that this is faster, the downside is that this may crash at runtime. Also, some special parts (for example the search index) do require everything to be loaded, which is what we account for here.

I'm still not decided on which approach is better long-term, a strong argument in favor of 1) is that it's conceptually simpler. Then again, a potential multi-second startup time is a bit painful as well. I've thought about adding a progress bar to ease the pain, but then again we don't want additional dependencies and we don't want our own progressbar implementation. 😛